### PR TITLE
Make the CUDA kernel of concat correct and fix unit tests.

### PR DESCRIPTION
**Pull request by qingqing01**
Fix https://github.com/PaddlePaddle/Paddle/issues/11540

### CODE REVIEW

File: paddle/fluid/operators/math/concat.cu [hunk 1]
```diff
@@ -22,43 +22,24 @@ namespace paddle {
...
   int tid_x = blockIdx.x * blockDim.x + threadIdx.x;
-  int segment = upper_bound<int>(input_cols, col_size, tid_x) - 1;
-
-  int curr_offset = input_cols[segment];
-  int curr_segment = segment;
   for (; tid_x < output_cols; tid_x += blockDim.x * gridDim.x) {
-    T curr_col_offset;
-    while ((curr_col_offset = input_cols[curr_segment + 1]) <= tid_x) {
+    int curr_segment = 0;
+    int curr_offset = input_cols[0];
```
Comment (chengduoZH) [inline paddle/fluid/operators/math/concat.cu]: It seems that line 31~32 can be out of loop.

**Comment by chengduoZH** [comment]
@qingqing01 
I know the reason what causes the illegal memory access, I use `T` as the type of [`curr_col_offset`](https://github.com/PaddlePaddle/Paddle/pull/11541/files#diff-0ee048f63368a2a61b90a9653d760d81L54), when the `T` is a float, and the input shape is too big to excess the representation of a float, it will cause the program to crash. It is my fault.

File: paddle/fluid/operators/math/concat.cu (hunk 1 above)
Comment (qingqing01) [reply paddle/fluid/operators/math/concat.cu]: Done. Thanks!

**Review by chengduoZH — APPROVED** [review]
LGTM

**Pull request merged**